vscode-extension/server/src/server.ts

```diff
@@ -151,8 +152,30 @@ connection.onDidChangeConfiguration(change => {
 
 documents.onDidClose(e => documentSettings.delete(e.document.uri));
 
-documents.onDidSave(async change => {
-  gatherAndReportDiagnostics(change.document);
+// when a document is changed or saved, we want to re-run diagnostics
+let changeToProcessNext: null | TextDocumentChangeEvent<TextDocument> = null;
+let processing = false;
+const processChange = async () => {
+  if (processing || !changeToProcessNext) return;
+  processing = true;
+  try {
+    const doc = changeToProcessNext.document;
+    changeToProcessNext = null;
+    await gatherAndReportDiagnostics(doc);
+  } finally {
+    processing = false;
+    if (changeToProcessNext) {
+      processChange();
+    }
+  }
+};
+documents.onDidSave(change => {
+  changeToProcessNext = change;
+  processChange();
+});
+documents.onDidChangeContent(change => {
+  changeToProcessNext = change;
+  processChange();
 });
```

Comment:
maybe this is fancier than it needs to be - in any case, it'll go away when we get rid of this server.ts in favor of a fully Dark-based LSP server